### PR TITLE
[1000 PR] Artifactory & Quay Perceivers added for OpsSight

### PR DESCRIPTION
**Pull request by gautambaghel**
woo - Tomas Gonzalez

quay perceiver and expose services made

artifactory dumper and debugger level for perceiver working

added go sum

basic rc and service made for artifactory controllers

artifactory dumper and debugger level for perceiver working

added go sum

### CODE REVIEW

**Comment by operator-r2d2** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: *<a href="https://github.com/blackducksoftware/synopsys-operator/pull/1000#" title="Author self-approved">gautambaghel</a>*
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please assign **msenmurugan**
You can assign the PR to them by writing `/assign @msenmurugan` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=blackducksoftware%2Fsynopsys-operator).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/blackducksoftware/synopsys-operator/blob/release-2019.10.x/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by yashbhutwala** [comment]
Yo!  Can you fix the CI issue and make a github issue with a description of what this PR solves (crossref to jira if need be)

File: go.sum [hunk 1]
```diff
@@ -44,8 +44,9 @@ github.com/beorn7/perks v1.0.1 h1:VlbKKnNfV8bJzeqoa4cOKqO6bYr3WgKZxO8Z16+hsOM=
 github.com/beorn7/perks v1.0.1/go.mod h1:G2ZrVWU2WbWT9wwq4/hrbKbnv/1ERSJQ0ibhJ6rlkpw=
 github.com/blackducksoftware/horizon v0.0.0-20190625151958-16cafa9109a3 h1:noI1RY2cUFZfdZMIz1+1LzT8ZeuWK703gwmH/ZC2YnQ=
 github.com/blackducksoftware/horizon v0.0.0-20190625151958-16cafa9109a3/go.mod h1:zprIpuTlP5utWbWF58V+/SOx9kbWm8PbXLUBwUvPh+0=
-github.com/blang/semver v3.5.0+incompatible/go.mod h1:kRBLl5iJ+tD4TcOOxsy/0fnwebNt5EWlYSAyrTnjyyk=
-github.com/cespare/xxhash v1.1.0/go.mod h1:XrSqR1VqqWfGrhpAt58auRo0WTKS1nRRg3ghfAqPWnc=
+github.com/blackducksoftware/perceivers v2.2.3+incompatible h1:xQwNY2d5WyysqinOLXS1Yz/BAlhfJ9zlU8sz+54f/h4=
```
Comment (msenmurugan) [inline go.sum]: why perceivers?

File: pkg/api/opssight/v1/types.go [hunk 2]
```diff
@@ -107,11 +108,15 @@ type PodPerceiver struct {
...
 // Perceiver stores the Perceiver configuration
 type Perceiver struct {
-	EnableImagePerceiver      bool          `json:"enableImagePerceiver"`
-	EnablePodPerceiver        bool          `json:"enablePodPerceiver"`
-	PodPerceiver              *PodPerceiver `json:"podPerceiver,omitempty"`
-	AnnotationIntervalSeconds int           `json:"annotationIntervalSeconds"`
-	DumpIntervalMinutes       int           `json:"dumpIntervalMinutes"`
+	EnableImagePerceiver             bool          `json:"enableImagePerceiver"`
+	EnableArtifactoryPerceiver       bool          `json:"enableArtifactoryPerceiver"`
+	EnableArtifactoryPerceiverDumper bool          `json:"enableArtifactoryPerceiverDumper"`
```
Comment (msenmurugan) [inline pkg/api/opssight/v1/types.go]: why 2 enables Artifactory Perceiver?

File: pkg/api/opssight/v1/types.go [hunk 3]
```diff
@@ -107,11 +108,15 @@ type PodPerceiver struct {
...
+	EnableQuayPerceiver              bool          `json:"enableQuayPerceiver"`
+	EnablePodPerceiver               bool          `json:"enablePodPerceiver"`
+	PodPerceiver                     *PodPerceiver `json:"podPerceiver,omitempty"`
+	AnnotationIntervalSeconds        int           `json:"annotationIntervalSeconds"`
+	DumpIntervalMinutes              int           `json:"dumpIntervalMinutes"`
+	Expose                           string        `json:"expose"`
```
Comment (msenmurugan) [inline pkg/api/opssight/v1/types.go]: what are we exposing for Perceivers?

File: pkg/opssight/ctl_opssight.go [hunk 4]
```diff
@@ -59,7 +59,12 @@ type CRSpecBuilderFromCobraFlags struct {
 	ScannerPodReplicaCount                          int
 	ScannerPodImageDirectory                        string
 	PerceiverEnableImagePerceiver                   string
+	PerceiverEnableArtifactoryPerceiver             string
+	PerceiverEnableArtifactoryPerceiverDumper       string
+	PerceiverEnableQuayPerceiver                    string
 	PerceiverEnablePodPerceiver                     string
+	PerceiverArtifactoryExpose                      string
+	PerceiverQuayExpose                             string
```
Comment (msenmurugan) [inline pkg/opssight/ctl_opssight.go]: Similar question, what are we exposing for Artifactory or Quay

File: pkg/opssight/ctl_opssight.go [hunk 5]
```diff
@@ -163,7 +168,20 @@ func (ctl *CRSpecBuilderFromCobraFlags) AddCRSpecFlagsToCommand(cmd *cobra.Comma
...
+	cmd.Flags().StringVar(&ctl.PerceiverEnableArtifactoryPerceiver, "enable-artifactory-processor", ctl.PerceiverEnableArtifactoryPerceiver, "If true, Artifactory Processor discovers artifactory images for scanning [true|false]")
+	cmd.Flags().StringVar(&ctl.PerceiverEnableArtifactoryPerceiverDumper, "enable-artifactory-processor-dumper", ctl.PerceiverEnableArtifactoryPerceiverDumper, "If true, Artifactory Processor dumps all docker images in an artifactory instance for scanning [true|false]")
+	cmd.Flags().StringVar(&ctl.PerceiverEnableQuayPerceiver, "enable-quay-processor", ctl.PerceiverEnableQuayPerceiver, "If true, Quay Processor discovers quay images for scanning [true|false]")
 	cmd.Flags().StringVar(&ctl.PerceiverEnablePodPerceiver, "enable-pod-processor", ctl.PerceiverEnablePodPerceiver, "If true, Pod Processor discovers pods for scanning [true|false]")
+	if master {
+		cmd.Flags().StringVar(&ctl.PerceiverArtifactoryExpose, "expose-artifactory-processor", util.NONE, "Type of service for Artifactory processor [NODEPORT|LOADBALANCER|OPENSHIFT|NONE]")
```
Comment (msenmurugan) [inline pkg/opssight/ctl_opssight.go]: why we need to have master?

File: pkg/opssight/ctl_opssight.go [hunk 6]
```diff
@@ -163,7 +168,20 @@ func (ctl *CRSpecBuilderFromCobraFlags) AddCRSpecFlagsToCommand(cmd *cobra.Comma
...
+	} else {
+		cmd.Flags().StringVar(&ctl.PerceiverArtifactoryExpose, "expose-artifactory-processor", ctl.PerceiverArtifactoryExpose, "Type of service for Artifactory processor [NODEPORT|LOADBALANCER|OPENSHIFT|NONE]")
+	}
+	if master {
+		cmd.Flags().StringVar(&ctl.PerceiverQuayExpose, "expose-quay-processor", util.NONE, "Type of service for Quay processor [NODEPORT|LOADBALANCER|OPENSHIFT|NONE]")
```
Comment (msenmurugan) [inline pkg/opssight/ctl_opssight.go]: why we need to have master?

File: pkg/opssight/perceivers.go [hunk 7]
```diff
@@ -63,6 +63,36 @@ func (p *SpecConfig) ImagePerceiverReplicationController() (*components.Replicat
...
 }
 
+// ArtifactoryPerceiverReplicationController creates a replication controller for the artifactory perceiver
+func (p *SpecConfig) ArtifactoryPerceiverReplicationController() (*components.ReplicationController, error) {
+	name := fmt.Sprintf("artifactory-perceiver")
+	image := fmt.Sprintf("artifactory-perceiver")
+
+	rc := p.perceiverReplicationController(name, 1)
+
+	pod, err := p.perceiverPod(name, image, util.GetResourceName(p.opssight.Name, util.OpsSightName, p.names["perceiver-service-account"]))
```
Comment (msenmurugan) [inline pkg/opssight/perceivers.go]: perceiver service account will have god permission? do we still need them?

File: pkg/opssight/perceivers.go [hunk 8]
```diff
@@ -63,6 +63,36 @@ func (p *SpecConfig) ImagePerceiverReplicationController() (*components.Replicat
...
+}
+
+// QuayPerceiverReplicationController creates a replication controller for the quay perceiver
+func (p *SpecConfig) QuayPerceiverReplicationController() (*components.ReplicationController, error) {
+	name := fmt.Sprintf("quay-perceiver")
+	image := fmt.Sprintf("quay-perceiver")
+
+	rc := p.perceiverReplicationController(name, 1)
+
+	pod, err := p.perceiverPod(name, image, util.GetResourceName(p.opssight.Name, util.OpsSightName, p.names["perceiver-service-account"]))
```
Comment (msenmurugan) [inline pkg/opssight/perceivers.go]: perceiver service account will have god permission? do we still need them?

File: pkg/opssight/perceivers.go [hunk 9]
```diff
@@ -144,7 +174,12 @@ func (p *SpecConfig) perceiverContainer(name string, image string) (*components.
 	if err != nil {
 		return nil, errors.Annotatef(err, "unable to add the volume mount to %s container", name)
 	}
-
+	if strings.Contains(name, "artifactory") || strings.Contains(name, "quay") {
+		if p.opssight.Spec.Perceiver.EnableArtifactoryPerceiverDumper {
+			// container.AddEnv()
```
Comment (msenmurugan) [inline pkg/opssight/perceivers.go]: Are you missing something??

File: pkg/opssight/prometheus.go [hunk 10]
```diff
@@ -215,6 +215,9 @@ func (p *SpecConfig) PerceptorMetricsConfigMap() (*components.ConfigMap, error)
 	if p.opssight.Spec.Perceiver.EnableImagePerceiver {
 		targets = append(targets, fmt.Sprintf("%s:%d", util.GetResourceName(p.opssight.Name, util.OpsSightName, p.names["image-perceiver"]), 3002))
 	}
+	if p.opssight.Spec.Perceiver.EnableArtifactoryPerceiver {
```
Comment (msenmurugan) [inline pkg/opssight/prometheus.go]: Enable Quay Perceiver metrics is missing

File: pkg/api/opssight/v1/types.go (hunk 2 above)
Comment (gautambaghel) [reply pkg/api/opssight/v1/types.go]: The second enables dumping all artifactory images into perceptor. Normally it'd work with webhooks, if something is pushed it adds to perceptor.

File: pkg/api/opssight/v1/types.go (hunk 3 above)
Comment (gautambaghel) [reply pkg/api/opssight/v1/types.go]: For Webhooks, they send a JSON to the perceivers with Image Info which needs to be scanned

File: pkg/opssight/ctl_opssight.go (hunk 4 above)
Comment (gautambaghel) [reply pkg/opssight/ctl_opssight.go]: For Webhooks, they send a JSON to the perceivers with Image Info which needs to be scanned

File: pkg/opssight/ctl_opssight.go (hunk 5 above)
Comment (gautambaghel) [reply pkg/opssight/ctl_opssight.go]: I kinda copied the exposing code of core/metrics. I'll change it.

File: pkg/opssight/ctl_opssight.go (hunk 6 above)
Comment (gautambaghel) [reply pkg/opssight/ctl_opssight.go]: I kinda copied the exposing code of core/metrics. I'll change it.

File: pkg/opssight/perceivers.go (hunk 7 above)
Comment (gautambaghel) [reply pkg/opssight/perceivers.go]: I don't think we do but I thought GetResourceName just gets the name from upstream. I can create my own replication controller/ pod is that what you're saying me to do?

File: pkg/opssight/perceivers.go (hunk 8 above)
Comment (gautambaghel) [reply pkg/opssight/perceivers.go]: I don't think we do but I thought GetResourceName just gets the name from upstream. I can create my own replication controller/ pod is that what you're saying me to do?

File: pkg/opssight/perceivers.go (hunk 9 above)
Comment (gautambaghel) [reply pkg/opssight/perceivers.go]: I'll remove this code

File: pkg/opssight/prometheus.go (hunk 10 above)
Comment (gautambaghel) [reply pkg/opssight/prometheus.go]: thanks, adding it now

File: go.sum (hunk 1 above)
Comment (gautambaghel) [reply go.sum]: removing :)

**Comment by yashbhutwala** [comment]
@msenmurugan @gautambaghel do we want this for 10.0 or 11.0?

**Pull request closed without merging**